tracer/events.go

```diff
@@ -125,29 +127,28 @@ func startPerfEventMonitor(ctx context.Context, perfEventMap *ebpf.Map,
 	}
 }
 
-// startPollingPerfEventMonitor spawns a goroutine that receives events from
-// the given perf event map by periodically polling the perf event buffer.
+// startTraceEventMonitor spawns a goroutine that receives trace events from
+// the kernel by periodically polling the underlying perf event buffer.
 // Events written to the perf event buffer do not wake user-land immediately.
 //
-// For each received event, triggerFunc is called. triggerFunc may NOT store
-// references into the buffer that it is given: the buffer is re-used across
-// calls. Returns a function that can be called to retrieve perf event array
+// Returns a function that can be called to retrieve perf event array
 // error counts.
-func startPollingPerfEventMonitor(ctx context.Context, perfEventMap *ebpf.Map,
-	pollFrequency time.Duration, perCPUBufferSize int, triggerFunc func([]byte, int),
-) func() (lost, noData, readError uint64) {
-	eventReader, err := perf.NewReader(perfEventMap, perCPUBufferSize)
+func (t *Tracer) startTraceEventMonitor(ctx context.Context,
+	traceOutChan chan<- *host.Trace) func() (lost, noData, readError uint64) {
+	eventReader, err := perf.NewReader(t.ebpfMaps["trace_events"],
+		t.samplesPerSecond*int(unsafe.Sizeof(C.Trace{})))
 	if err != nil {
-		log.Fatalf("Failed to setup perf reporting via %s: %v", perfEventMap, err)
+		log.Fatalf("Failed to setup perf reporting via %s: %v", t.ebpfMaps["trace_events"], err)
 	}
 
 	// A deadline of zero is treated as "no deadline". A deadline in the past
 	// means "always return immediately". We thus set a deadline 1 second after
 	// unix epoch to always ensure the latter behavior.
 	eventReader.SetDeadline(time.Unix(1, 0))
 
-	pollTicker := time.NewTicker(pollFrequency)
+	pollTicker := time.NewTicker(t.intervals.TracePollInterval())
```

Comment:
To prevent leaks and have the right scoping, `pollTicker`could be moved inside the Go routine and add `defer pollTicker.Stop()`.